documents/extension/funding/service.go

```diff
@@ -0,0 +1,176 @@
+package funding
+
+import (
+	"context"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/centrifuge/go-centrifuge/utils"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/centrifuge/go-centrifuge/centerrors"
+	"github.com/centrifuge/go-centrifuge/documents"
+	"github.com/centrifuge/go-centrifuge/errors"
+	clientfundingpb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/funding"
+)
+
+// Service defines specific functions for extension funding
+type Service interface {
+	documents.Service
+	DeriveFromPayload(ctx context.Context, req *clientfundingpb.FundingCreatePayload, identifier []byte) (documents.Model, error)
+	DeriveFundingResponse(model documents.Model, payload *clientfundingpb.FundingCreatePayload) (*clientfundingpb.FundingResponse, error)
+}
+
+// service implements Service and handles all funding related persistence and validations
+type service struct {
+	documents.Service
+	tokenRegFinder func() documents.TokenRegistry
+}
+
+const fundingLabel = "centrifuge_funding"
+const fundingFieldKey = "centrifuge_funding[{IDX}]."
+const fundingIdx = "{IDX}"
+
+// DefaultService returns the default implementation of the service.
+func DefaultService(
+	srv documents.Service,
+	tokenRegFinder func() documents.TokenRegistry,
+) Service {
+	return service{
+		Service:        srv,
+		tokenRegFinder: tokenRegFinder,
+	}
+}
+
+func generateKey(idx, fieldName string) string {
+	return strings.Replace(fundingFieldKey, fundingIdx, idx, -1) + fieldName
+}
+
+func keyFromJsonTag(idx, jsonTag string) (key string, err error) {
+	correctJsonParts := 2
+	jsonKeyIdx := 0
+
+	// example `json:"days,omitempty"`
+	jsonParts := strings.Split(jsonTag, ",")
+	if len(jsonParts) == correctJsonParts {
+		return generateKey(idx, jsonParts[jsonKeyIdx]), nil
+
+	}
+	return key, ErrNoFundingField
+
+}
+
+func defineFundingIdx(model documents.Model) (attr documents.Attribute, err error) {
+	key, err := documents.AttrKeyFromLabel(fundingLabel)
+	if err != nil {
+		return attr, err
+	}
+
+	if !model.AttributeExists(key) {
+		return documents.NewAttribute(fundingLabel, documents.AttrString, "0")
```

Comment:
@vedhavyas first I wanted it to define as int256 but it is not so easy to compare `big.Int`

Reply:
Not everything is String type right? There will be decimals, integers etc...

Reply:
Agree I will change it to int256